ouroboros-network/test/CDDL.hs

```diff
@@ -3,134 +3,215 @@
-type BF = BlockFetch DummyBytes
-type HS = Handshake Word32 CBOR.Term
+-- These files must be in place:
+specFile :: FilePath
+specFile = "test/messages.cddl"
+
+cddlTool :: FilePath
+cddlTool = "cddl"
+
+diag2cborTool :: FilePath
+diag2cborTool = "diag2cbor.rb"
 
 main :: IO ()
-main = generateAndDecode "cddl" "diag2cbor.rb" "test/messages.cddl" 100
+main = defaultMain tests
+
+tests :: TestTree
+tests =
+  adjustOption (const $ QuickCheckMaxSize 20) $ testGroup "messages.cddl-spec"
+  [
+-- These tests call the CDDL-tool to parse an arbitray message.
+-- The parser of the CDDL-tool is slow (exponential runtime & space).
+    testProperty "encode ChainSync"             prop_specCS
+  , testProperty "encode Request-Response"      prop_specReqResp
+  , testProperty "encode BlockFetch"            prop_specBF
+  , testProperty "encode TxSubmission"          prop_specTxSubmission
+  , testProperty "encode Handshake"             prop_specHandshake
+
+-- Test the parsers with CDDL-generated messages.
+  , testProperty "generate and decode" $ ioProperty $ generateAndDecode 100 specFile
+  ]
+
+-- The concrete/monomorphic types used for the test.
+type MonoCodec x = Codec x Codec.CBOR.Read.DeserialiseFailure IO ByteString
+type CS = ChainSync BlockHeader (Point BlockHeader, BlockNo)
+type RR = ReqResp DummyBytes DummyBytes
+type BF = BlockFetch Block
+type HS = Handshake VersionNumber CBOR.Term
+type TS = TxSubmission TxId Tx
+
+codecCS :: MonoCodec CS
+codecCS = codecChainSync Serialise.encode (fmap const Serialise.decode)
+                         Serialise.encode (Serialise.decode :: CBOR.Decoder s (Point BlockHeader))
+                         Serialise.encode (Serialise.decode :: CBOR.Decoder s (Point BlockHeader, BlockNo))
+
+codecBF :: MonoCodec BF
+codecBF = codecBlockFetch Serialise.encode (fmap const Serialise.decode) Serialise.encode Serialise.decode
+
+codecTS :: MonoCodec TS
+codecTS = codecTxSubmission Serialise.encode Serialise.decode Serialise.encode Serialise.decode
+
+prop_specCS :: AnyMessageAndAgency CS -> Property
+prop_specCS = prop_CDDLSpec (0, codecCS)
+
+prop_specReqResp :: AnyMessageAndAgency RR -> Property
+prop_specReqResp = prop_CDDLSpec (1, codecReqResp)
 
-generateAndDecode :: FilePath -> FilePath -> FilePath -> Int -> IO ()
-generateAndDecode cddlCmd diag2cborCmd cddlSpec rounds = do
-    diags <- generateCBORDiag cddlCmd cddlSpec rounds
-    forM_ diags $ \diag -> do
+prop_specBF :: AnyMessageAndAgency BF -> Property
+prop_specBF = prop_CDDLSpec (3, codecBF)
+
+prop_specTxSubmission :: AnyMessageAndAgency TS -> Property
+prop_specTxSubmission = prop_CDDLSpec (4, codecTS)
+
+prop_specHandshake :: AnyMessageAndAgency HS -> Property
+prop_specHandshake = prop_CDDLSpec (5, codecHandshake)
+
+prop_CDDLSpec :: (Word, MonoCodec ps) -> AnyMessageAndAgency ps -> Property
+prop_CDDLSpec (tagWord, codec) (AnyMessageAndAgency agency msg)
+    = ioProperty $ do
+         print $ BSL.unpack wrappedMsg
```

Comment:
Don't print out raw messages to `stdout`, hide it in a comment if you need it for debugging or add a `cddl-verbose` flag.